adoptingerlang.org

```diff
@@ -3405,27 +3542,269 @@ $ kubectl apply -k deployment/overlays/dev
 $ microk8s.enable registry
 $ microk8s.enable dns
 #+END_SRC
 
-=/etc/docker/daemon.json=
+To allow the host Docker to publish to this registry it must be added to =/etc/docker/daemon.json=:
 
 #+BEGIN_SRC json
 {
-  ...
+  [...]
   "insecure-registries" : ["localhost:32000"]
-  ...
+  [...]
 }
 #+END_SRC
 
-*** Database Migrations
-**** Jobs
-**** Init Containers
-*** StatefulSet
+Tilt works off of a file named =Tiltfile= at the root of the project. Stepping through the =Tiltfile= at the root of [[https://github.com/adoptingerlang/service_discovery][service_discovery]] it begins with:
+
+#+BEGIN_SRC python
+allow_k8s_contexts("microk8s")
+default_registry('127.0.0.1:32000')
+#+END_SRC
+
+=allow_k8s_contexts= restricts the Kubernetes context that Tilt will run against to =microk8s=, this is to protect the user from accidentally deploying their local development resources and images to an external Kubernetes cluster, like production! =default_registry= sets where to push Docker images built by Tilt, in this case the registry enabled in =microk8s= is used.
+
+Next the build of a couple of Docker images is configured, starting with =service_discovery_sql=:
+
+#+BEGIN_SRC python
+custom_build(
+    'service_discovery_sql',
+    'docker buildx build -o type=docker --target dev_sql --tag $EXPECTED_REF .',
+    ['apps/service_discovery_postgres/priv/migrations'],
+    entrypoint="cp /app/sql/* /flyway/sql"
+)
+#+END_SRC
+
+This image is built from the =Dockerfile= target =dev_sql=:
+
+#+BEGIN_SRC dockerfile
+FROM busybox as dev_sql
+
+COPY apps/service_discovery_postgres/priv/migrations/ /app/sql/
+#+END_SRC
+
+It only contains the SQL migration files and uses =busybox= so that the =entrypoint= can use =cp=. The third argument to =custom_build=, =['apps/service_discovery_postgres/priv/migrations']=, tells Tilt to rebuild the image if any file in that directory, in this case the SQL migration files, changes. We will see how this image is used when we get to the kustomize deployment later in the =Tiltfile=.
+
+The next image built is =service_discovery=, the main image used by the =Deployment=:
+
+#+BEGIN_SRC python
+custom_build(
+    'service_discovery',
+    'docker buildx build -o type=docker --target dev_release --tag $EXPECTED_REF .',
+    ['.'],
+    live_update=[
+        sync('rebar.config', '/app/src/rebar.config'),
+        sync('apps', '/app/src/apps'),
+        run('rebar3 as tilt compile'),
+        run('/app/_build/tilt/rel/service_discovery/bin/service_discovery restart')
+    ],
+    ignore=["rebar.lock", "apps/service_discovery_postgres/priv/migrations/"]
+)
+#+END_SRC
+
+Note that =rebar.lock= is explicitly ignored, this is because it gets rewritten at times when it hasn't actually changed and Tilt does no comparison to verify a change has actually occurred, so would needlessly run the =live_update= instructions when Rebar3 is run locally. The migration files are also ignored because that is handled by the other Docker image.
```

Comment:
we should figure that out, because there's no reason to rewrite the lock file if it hasn't changed. I thought all deps were sorted for minimal disruption as well.

Reply:
They are sorted, but it still is writing the file for no good reason.

Reply:
Made a PR https://github.com/erlang/rebar3/pull/2214